### PR TITLE
Move all build installer tasks to Gradle

### PR DESCRIPTION
**Pull request by ssoloff**
Currently, the installers created by a Travis build differ from those created during a local build because some tasks are performed exclusively as part of the Travis build script.  This PR moves those tasks into the Gradle build so a local build will produce the same installers as Travis.

#### Changes

* The existing `setRunnableLinux` task was renamed to `prepareInstallers`, and the Windows installers are now renamed as part of this task.
* Updating the engine version in `game_engine.properties` is now done as part of the new `prepareRootFiles` task.  In order to keep the build idempotent, the updated `game_engine.properties` file is written to `build/rootFiles/game_engine.properties`.  The installers and `all_platforms` zip archive now use that file instead of the original file in the repo root.
* As the `build_installers` Travis script did nothing more than run the `release` task after these changes, it was removed and its content inlined directly into `.travis.yml`.

#### Testing

I tested [building](https://travis-ci.org/ssoloff/triplea-game-triplea/builds/251616270) the installers and deploying a [release](https://github.com/ssoloff/triplea-game-triplea/releases/tag/1.9.0.0.698) on my fork.  I verified that the `game_engine.properties` file contained the correct value for `engine_version` when installing using the following artifacts:

* `all_platforms` zip archive
* Linux installer
* Windows x64 installer

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -153,7 +156,17 @@ task renameShadowJar(type: Copy, group: 'release', dependsOn: [shadowJar]) {
 task updateAssets(dependsOn: ['gitCloneAssets', 'gitPullAssets']) {
 }
 
-task allPlatform(type: Zip, group: 'release', dependsOn: [renameShadowJar, updateAssets]) {
+task prepareGameEngineProperties(type: Copy, group: 'release') {
+    inputs.file gameEnginePropertiesFile
+
+    from gameEnginePropertiesFile
+    filter { it.replaceAll(/^(engine_version)\s*=.*$/, "\$1 = $version") }
```
Comment (RoiEXLab) [inline build.gradle]: Wouldn't it be better to use the ant.propertyfile task to edit the properties file?

File: build.install4j [hunk 2]
```diff
@@ -40,7 +40,7 @@
       <dirEntry mountPoint="25" file="old" overwriteMode="4" shared="false" fileMode="644" uninstallMode="0" overrideFileMode="false" overrideOverwriteMode="false" overrideUninstallMode="false" entryMode="direct" subDirectory="old" excludeSuffixes="" dirMode="755" overrideDirMode="false">
         <exclude />
       </dirEntry>
-      <fileEntry mountPoint="22" file="./game_engine.properties" overwriteMode="4" shared="false" fileMode="644" uninstallMode="0" overrideFileMode="false" overrideOverwriteMode="false" overrideUninstallMode="false" />
+      <fileEntry mountPoint="22" file="./build/rootFiles/game_engine.properties" overwriteMode="4" shared="false" fileMode="644" uninstallMode="0" overrideFileMode="false" overrideOverwriteMode="false" overrideUninstallMode="false" />
```
Comment (RoiEXLab) [inline build.install4j]: Good idea to temporarily copy the file!

File: build.gradle [hunk 3]
```diff
@@ -203,14 +216,20 @@ task generateInstallers(type: Install4jTask, dependsOn: [renameShadowJar, copyJR
...
 }
 
-task setRunnableLinux(dependsOn: [generateInstallers]) {
+task prepareInstallers(group: 'release', dependsOn: [generateInstallers]) {
     doLast {
-        def linuxFile = "TripleA_${version}_unix.sh"
-        ant.chmod(dir: "$buildDir/releases", perm: '+x', includes: linuxFile)
+        def releasesDir = "$buildDir/releases"
+
+        ant.chmod(dir: releasesDir, perm: '+x', includes: "TripleA_${version}_unix.sh")
```
Comment (RoiEXLab) [inline build.gradle]: I'm wondering if this line is actually necessary.
Could you check if install4j generates files and marks them as executable?
Once they are uploaded the permission are reset anyways so the one who could benefit from this line is the developer.

File: build.gradle (hunk 1 above)
Comment (ssoloff) [inline build.gradle]: c7df933

It ends up being more verbose, but I agree with you that it is better in the long run.  Primarily because it's more readable, but secondarily because it would probably be necessary anyway if we ever have to replace multiple properties (because we use real values instead of tokens like `@version@` 😞).

The only issue I came across is that the Ant task rewrites portions of the file.  For example, after running Gradle, the output file has the following differences:

```diff
diff --git a/game_engine.properties b/build/rootFiles/game_engine.properties
index a7f87d1..c51c7ee 100644
--- a/game_engine.properties
+++ b/build/rootFiles/game_engine.properties
@@ -1,18 +1,19 @@
+#Sun, 09 Jul 2017 16:37:43 -0400
 ## Main configuration file for TripleA game client.
 ## Values here will usually not need to be adjusted.
 ## If updating this file, remember to save a backup
 ## first.
 
-engine_version = 1.9.0.0
+engine_version=1.9.0.0.7777
 
 ## URL where we will download the list of available maps
-map_list_file = https://raw.githubusercontent.com/triplea-game/triplea/master/triplea_maps.yaml?raw=true
+map_list_file=https\://raw.githubusercontent.com/triplea-game/triplea/master/triplea_maps.yaml?raw\=true
 ## Example that would use a local file config file, useful for map building or development
 # map_list_file = triplea_maps.yaml
 
 
 ## Location where we download latest lobby description file, this tells us where the latest lobby is located
-lobby_properties_file_url = https://raw.githubusercontent.com/triplea-game/triplea/master/lobby_server.yaml?raw=true
+lobby_properties_file_url=https\://raw.githubusercontent.com/triplea-game/triplea/master/lobby_server.yaml?raw\=true
 
 ## If there is any trouble downloading the latest lobby location, this backup will be used
-lobby_backup_url = 45.79.144.53:3304
+lobby_backup_url=45.79.144.53\:3304
```

That is, it removes non-essential whitespace and escapes characters per the properties file specification.  However, I don't think this is a problem.  I tested one of the generated installers to make sure I had no issues downloading the map list file nor connecting to the lobby using the rewritten URLs.

File: build.gradle (hunk 3 above)
Comment (ssoloff) [inline build.gradle]: Yes, install4j (at least on Linux) does not automatically set the executable bit for the Unix installer.

Actually, I interpreted your suggestion as checking to see if install4j provided an option to set the Unix mode of the generated media file.  Unfortunately, while install4j does provide the ability to set the Unix mode for the installed files themselves, it does not appear to support this feature for the actual installer.

However, I found an option that allows us to override the media file name.  So I replaced the Gradle code that renamed the Windows installers with custom media file names in the install4j build script (see the `mediaFileName` attribute on the Windows 32/64 children of the `mediaSets` element).  So, in the end, it was worth the investigation. :smile:  bd000e6

File: build.gradle [hunk 4]
```diff
@@ -153,7 +158,24 @@ task renameShadowJar(type: Copy, group: 'release', dependsOn: [shadowJar]) {
...
+            from gameEnginePropertiesFile
+            into gameEnginePropertiesArtifactFile.parent
+        }
+        ant.propertyfile(file: gameEnginePropertiesArtifactFile) {
+            entry key: 'engine_version', value: version
+        }
+    }
+}
+
+task prepareRootFiles(group: 'release', dependsOn: [prepareGameEngineProperties]) {}
```
Comment (RoiEXLab) [inline build.gradle]: Does it make sense to have a task require a single other task without any action itself?

File: build.gradle (hunk 3 above)
Comment (RoiEXLab) [inline build.gradle]: That's pretty neat!
I knew the installer name scheme could be changed (I did this myself in a PR a couple days ago) bit I didn't know this name scheme could be changed for the individual installers

File: build.gradle (hunk 4 above)
Comment (ssoloff) [inline build.gradle]: I think I started working top-down and just assumed I would do everything root-file-related in `prepareRootFiles`.  Then when I made the task incremental, it didn't make sense anymore, and I broke out `prepareGameEngineProperties` thinking there could be more root file processing in the future.  Since there isn't any now, I went ahead and removed `prepareRootFiles`.  We can always add it back if a second root file comes along. :smile: bafbc8c

File: .travis.yml [hunk 5]
```diff
@@ -21,7 +21,7 @@ before_deploy:
 - ./.travis/install_install4j
 - ENGINE_VERSION=$(grep "engine_version" game_engine.properties | sed 's/.*= *//g')
 - TAGGED_VERSION="$ENGINE_VERSION.$TRAVIS_BUILD_NUMBER"
-- ./.travis/build_installers $TAGGED_VERSION
+- TAGGED_VERSION=$TAGGED_VERSION ./gradlew release
```
Comment (DanVanAtta) [inline .travis.yml]: I don't quite follow here in this assignment to `TAGGED-VERSION`. Could you the clarify what is being done?

File: build.gradle [hunk 6]
```diff
@@ -153,7 +158,22 @@ task renameShadowJar(type: Copy, group: 'release', dependsOn: [shadowJar]) {
 task updateAssets(dependsOn: ['gitCloneAssets', 'gitPullAssets']) {
 }
 
-task allPlatform(type: Zip, group: 'release', dependsOn: [renameShadowJar, updateAssets]) {
+task prepareGameEngineProperties(group: 'release') {
```
Comment (DanVanAtta) [inline build.gradle]: A comment here with some context on why we are doing this operation would perhaps be very useful for current/future reference. I don't quite rightly know offhand why this copy would happen.

File: .travis.yml (hunk 5 above)
Comment (ssoloff) [inline .travis.yml]: Sure...  Our Gradle build script has a data dependency on the `TAGGED_VERSION` environment variable (`build.gradle:14`).  Because `TAGGED_VERSION` is not exported on the previous line, it must explicitly be defined to appear in the environment of the Gradle process.

I think we used to export it when everything was in `.travis.yml`.  But, IIRC, when you broke the different concerns of `.travis.yml` into separate shell scripts (f687bc9), you made it a point to explicitly show the data dependencies (a Good Thing, IMO).  So I did it this way rather than prefixing the previous line with an `export`.

(Going through the history of `build_installers`, I see that I had to fix a build break in 4785d02 by exporting `TAGGED_VERSION` within the context of that script only.  I think I said in the PR at the time that the `TAGGED_VERSION=$TAGGED_VERSION` prefix looked messy, and I went with exporting `TAGGED_VERSION` instead.  As I said above, I didn't think that was an option here so we could keep the data dependencies explicit.)

Personally, I would prefer if we passed this value as a Gradle project variable (e.g. `./gradlew release -Pversion=$TAGGED_VERSION`), but I didn't want to do too much in this PR.

Comment (DanVanAtta) [inline .travis.yml]: thanks. Is there missing backticks on `./gradlew release`? It seems like this would assign `TAGGED_VERSION` to itself plus the string value "./gradlew release".

Comment (ssoloff) [inline .travis.yml]: Hmm, I don't think they are required.  It's working for me both locally (bash) and on [Travis](https://travis-ci.org/ssoloff/triplea-game-triplea/builds/251834683#L4438) (dash).

I can't find a Bourne shell reference to give you to prove this, but I tried a little experiment in my bash shell:

```shell
$ node -e "console.log(process.env['FOO'])"
undefined
$ FOO=bar node -e "console.log(process.env['FOO'])"
bar
$ FOO=bar baz node -e "console.log(process.env['FOO'])"
bash: baz: command not found...
$ FOO="bar baz" node -e "console.log(process.env['FOO'])"
bar baz
```

It looks like the shell stops parsing the environment variable value when it encounters whitespace.  So, technically, we could surround `$TAGGED_VERSION` with quotes just in case it contains whitespace, but it shouldn't be necessary for the command itself:

```shell
TAGGED_VERSION="$TAGGED_VERSION" ./gradlew release
```

And please note that this is **not** overwriting the existing `TAGGED_VERSION` environment variable in the parent shell.  It is simply defining a new variable (with the same name) in the child process.  For example:

```shell
$ echo $FOO

$ FOO=bar
$ echo $FOO
bar
$ FOO=baz node -e "console.log(process.env['FOO'])"
baz
$ echo $FOO
bar
```

Comment (DanVanAtta) [inline .travis.yml]: Thank you, question, could we simply write the two statements on different lines, or is it on same line due to variable scoping? Sorry, it's an odd syntax for me to see, hence some confusion.

File: build.gradle (hunk 6 above)
Comment (ssoloff) [inline build.gradle]: I added a task description in ddb0dcc.  Please let me know if it needs further clarification.

File: .travis.yml (hunk 5 above)
Comment (ssoloff) [inline .travis.yml]: Unfortunately, no.  AFAIK, the child process environment variable definition must appear at the beginning of the command that starts the child process.

I really wish I could find you an "official" reference for this feature for bash or any other Bourne-style shell, but I can't locate one at the moment.  The best I can do is quote you from the bash man page:

> The environment for any simple command or function may be augmented temporarily by prefixing it with parameter assignments, as described above in PARAMETERS.  These assignment statements affect only the environment seen by that command.

This [SE question](https://unix.stackexchange.com/questions/130985/if-processes-inherit-the-parents-environment-why-do-we-need-export) covers this feature, as well as some general theory.

Basically, this line:

```
TAGGED_VERSION="$ENGINE_VERSION.$TRAVIS_BUILD_NUMBER"
```

is defining a _shell variable_ named `TAGGED_VERSION`.  Then this line:

```
TAGGED_VERSION=$TAGGED_VERSION ./gradlew release
```

is creating an _environment variable_ in the child process named `TAGGED_VERSION` whose value is taken from the _shell variable_ `TAGGED_VERSION`.

Comment (RoiEXLab) [inline .travis.yml]: One thing I'm still wondering about: Why is it even necessary to define a "child variable" if the actual environment variable is already set to the correct value?

Comment (RoiEXLab) [inline .travis.yml]: In other words: wouldn't it just work if we exported the TAGGED_VERSION var when it's being defined in the travis script?

Comment (ssoloff) [inline .travis.yml]: > Why is it even necessary to define a "child variable" if the actual environment variable is already set to the correct value?

Technically, there is no _environment variable_ named `TAGGED_VERSION` in the parent script; there is only a _shell variable_ of that name.

> wouldn't it just work if we exported the TAGGED_VERSION var when it's being defined in the travis script?

Yes, exporting `TAGGED_VERSION` would be functionally equivalent and is what was done prior to f687bc9.  However, as mentioned above, there was a conscious decision in f687bc9 to make the data dependencies explicit.  From that commit's message:

> Make helper scripts take parameters to show data dependency between travis.yml script and helpers

In fact, it was the implicit dependency of `build.gradle` on the `TAGGED_VERSION` environment variable that led to the build break caused by f687bc9 and fixed in 4785d02.

Since there is still some discomfort around this syntax, I will push a commit that changes `build.gradle` to not depend on an environment variable but to always use a Gradle project variable instead.

Comment (ssoloff) [inline .travis.yml]: I changed the build script to use a Gradle project property instead of an environment variable to define the version.  However, while working through this exercise, a few other issues came up.  They are minor and easily resolved, but it may require a team discussion to decide the best path forward.

Therefore, I'm going to propose that the change to a Gradle project property be postponed to a separate PR.

Comment (DanVanAtta) [inline .travis.yml]: Informative, thank you - not too often that shell vs environment comes up.
